app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenProviderImpl.kt

```diff
@@ -48,56 +49,52 @@ object PoTokenProviderImpl : PoTokenProvider {
      * case the current [webPoTokenGenerator] threw an error last time
      * [PoTokenGenerator.generatePoToken] was called
      */
-    private fun getWebClientPoToken(videoId: String, forceRecreate: Boolean): PoTokenResult {
+    private suspend fun getWebClientPoToken(videoId: String, forceRecreate: Boolean): PoTokenResult {
         // just a helper class since Kotlin does not have builtin support for 4-tuples
         data class Quadruple<T1, T2, T3, T4>(val t1: T1, val t2: T2, val t3: T3, val t4: T4)
 
         val (poTokenGenerator, visitorData, streamingPot, hasBeenRecreated) =
-            synchronized(WebPoTokenGenLock) {
-                val shouldRecreate = webPoTokenGenerator == null || forceRecreate ||
-                    webPoTokenGenerator!!.isExpired()
+            webPoTokenGenLock.withLock {
+                val gen = webPoTokenGenerator
+                val shouldRecreate = forceRecreate || gen == null || gen.isExpired
 
                 if (shouldRecreate) {
-
-                    val innertubeClientRequestInfo = InnertubeClientRequestInfo.ofWebClient()
-                    innertubeClientRequestInfo.clientInfo.clientVersion =
-                        YoutubeParsingHelper.getClientVersion()
-
-                    webPoTokenVisitorData = YoutubeParsingHelper.getVisitorDataFromInnertube(
-                        innertubeClientRequestInfo,
-                        NewPipe.getPreferredLocalization(),
-                        NewPipe.getPreferredContentCountry(),
-                        YoutubeParsingHelper.getYouTubeHeaders(),
-                        YoutubeParsingHelper.YOUTUBEI_V1_URL,
-                        null,
-                        false
-                    )
-                    // close the current webPoTokenGenerator on the main thread
-                    webPoTokenGenerator?.let { Handler(Looper.getMainLooper()).post { it.close() } }
+                    webPoTokenVisitorData = withContext(Dispatchers.IO) {
+                        val innertubeClientRequestInfo = InnertubeClientRequestInfo.ofWebClient()
```

Comment:
Do I read this correctly, before it was executed blocking on the main thread and now it’s dispatched to the IO thread?